src/daft-plan/src/physical_ops/project.rs

```diff
@@ -63,37 +64,41 @@ impl Project {
                 }
 
                 // Then, see if we can fully translate the partition spec.
-                let maybe_new_pspec = input_pspec
-                    .by
-                    .as_ref()
-                    .unwrap()
+                let maybe_new_clustering_spec = by
                     .iter()
-                    .map(|e| Self::translate_partition_spec_expr(e, &old_colname_to_new_colname))
+                    .map(|e| Self::translate_clustering_spec_expr(e, &old_colname_to_new_colname))
                     .collect::<std::result::Result<Vec<_>, _>>();
-                maybe_new_pspec.map_or_else(
+                maybe_new_clustering_spec.map_or_else(
                     |()| {
-                        PartitionSpec::new(
-                            PartitionSchemeConfig::Unknown(Default::default()),
-                            input_pspec.num_partitions,
-                            None,
-                        )
+                        ClusteringSpec::Unknown(UnknownClusteringConfig::new(
+                            input_clustering_spec.num_partitions(),
+                        ))
                         .into()
                     },
-                    |new_pspec: Vec<Expr>| {
-                        PartitionSpec::new(
-                            input_pspec.scheme_config.clone(),
-                            input_pspec.num_partitions,
-                            Some(new_pspec),
+                    |new_clustering_spec: Vec<Expr>| match input_clustering_spec.as_ref() {
+                        Range(RangeClusteringConfig {
+                            num_partitions,
+                            descending,
+                            ..
+                        }) => ClusteringSpec::Range(RangeClusteringConfig::new(
+                            *num_partitions,
+                            new_clustering_spec,
+                            descending.clone(),
+                        ))
+                        .into(),
+                        Hash(HashClusteringConfig { num_partitions, .. }) => ClusteringSpec::Hash(
+                            HashClusteringConfig::new(*num_partitions, new_clustering_spec),
                         )
-                        .into()
+                        .into(),
+                        _ => unreachable!(),
```

Comment:
Feels a little odd to me that we need `unreachable!()` here because of the repeated calling of the `match input_clustering_spec.as_ref()` matches...

Only way I could think of is to refactor `let mut old_colname_to_new_colname = ...` up till `let maybe_new_clustering_spec = ...` as a separate function, and we can then call them in the outer `match input_clustering_spec.as_ref()` arm?

Reply:
Agreed, going to leave this for now and do a larger refactor later when we port this to use the `TreeVisitor` implementation for `Expr`.